### ISSUE TITLE
Create struct to encapsulate all rollbackable fields

### ISSUE THREAD

**Issue by MarcosNicolau** [lambdaclass/era_vm#196]
I’ve been working on implementing pubdata and refunds(see [here](https://github.com/lambdaclass/era_vm/pull/195)) to fix several integration tests for zksync-era with our VM. This requires adding several new fields (I believe I am even missing some):
```rust
pubdata: i32
pubdata_costs: Vec<i32>
refunds: Vec<i32>
read_storage_keys: Vec<StorageKey>
write_storage_keys: Vec<StorageKey>
```

All these fields need to be rollbackable. So far, only `state_storage` and `transient_storage` have been rollbacked, and so we were managing them by passing a snapshot to new frames, and when reverting we would just set the `state_storage` and `transient_storage` to their respective snapshot. 

The problem is, with all this new fields, we’d be passing a significant number of parameters to each opcode and frame. Additionally, whenever we need to revert a change, we’d have to update all relevant places to include each new field.

Before adding new features, I think it’s best to refactor the code by creating a struct that encapsulates all these fields and includes a rollback method that reverts each field, and a snapshot method that creates a copy of each field. This struct would be passed to each frame via the snapshot method, and in the case of any `ret.revert` or `ret.panic`, a rollback would be performed. This way, the opcodes would receive:

```rust
exec_opcode(&mut vm.state, &opcode, &mut rollbackable_struct)
```

This will also require an update to the integration on zksync-era, but it’s better to do it now while there are only a few fields rather than waiting until there are many more.

Additionally, there are other fields we should consider making rollbackable:
- events
- l2_l1_logs

**Comment by MarcosNicolau** [comment]
Btw, they are doing something similar but with a `WorldDiff` struct(see [here](https://github.com/matter-labs/vm2/blob/master/src/world_diff.rs)). What I am proposing is much more simpler, without all the methods they have(most of them are only used once across the whole codebase), only a struct with similar fields but whose only responsibility would be to rollback each field and create snapshots.

**Comment by juan518munoz** [comment]
I believe that an initial implementation akin to `WorldDiff` would be a good idea.

- `state_storage` and `transient_storage` must go inside of it.
- Implement the struct (and it's fields) with at least one test that makes use of it, we shouldn't push a struct with unused fields just in case, let's keep it as simple as possible.

One question:
```rs
exec_opcode(&mut vm.state, &opcode, &mut rollbackable_struct)
```
What's the point of passing a `rollbackable_struct` as an argument here? Wouldn't the `Callframes` already have a hold of them?

**Comment by MarcosNicolau** [comment]
Noted! Regarding your question: that is the struct that would get mutated and then if a `ret.panic` or `ret.revert` occurs, then we would set to the `vm.current_frame` stored snapshot.

**Comment by MarcosNicolau** [comment]
This has been solved in #197
